connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala

```diff
@@ -58,14 +84,33 @@ object functions {
    * @param messageName
    *   the protobuf MessageName to look for in descriptor file.
    * @param descFilePath
-   *   the protobuf descriptor file.
+   *   The Protobuf descriptor file. This file is usually created using `protoc` with
+   *   `--descriptor_set_out` and `--include_imports` options.
    * @since 3.4.0
    */
   @Experimental
   def from_protobuf(data: Column, messageName: String, descFilePath: String): Column = {
-    new Column(ProtobufDataToCatalyst(data.expr, messageName, descFilePath = Some(descFilePath)))
-    // TODO: Add an option for user to provide descriptor file content as a buffer. This
-    //       gives flexibility in how the content is fetched.
+    val fileContent = ProtobufUtils.readDescriptorFileContent(descFilePath)
+    new Column(ProtobufDataToCatalyst(data.expr, messageName, Some(fileContent)))
+  }
+
+  /**
+   * Converts a binary column of Protobuf format into its corresponding catalyst value.The
+   * Protobuf definition is provided through Protobuf `FileDescriptorSet`.
+   *
+   * @param data
+   *   the binary column.
+   * @param messageName
+   *   the protobuf MessageName to look for in the descriptor set.
+   * @param fileDescriptorSetBytes
+   *   Serialized Protobuf descriptor (`FileDescriptorSet`). Typically contents of file created
+   *   using `protoc` with `--descriptor_set_out` and `--include_imports` options.
+   * @since 3.5.0
+   */
+  @Experimental
+  def from_protobuf(data: Column, messageName: String, fileDescriptorSetBytes: Array[Byte])
```

Comment:
This is an example new API that takes byte array as the argument. 
I will send a separate PR for Python API.